### PR TITLE
Improvements to `Composed` and  `TruncatedBijector` fixing #40

### PR DESCRIPTION
**Pull request by torfjelde**
# Overview
This PR introduces most of the _performance_ improvements to `Composed` discussed in #46 and the `TruncatedBijector` discussed in #40 making `bijector(d::Distribution)` type-stable and providing a good "default" bijector.

In addition to the above, `Test.@inferred` is now used extensively in `test/interface.jl` to try and catch instability for the `<:Tuple` versions of `Composed` and `Stacked`.

Finally the project structure has been changed a bit (for the better). Implementations of `Bijector` are now found in the `src/bijectors` directory, e.g. `src/bijectors/scale.jl`. This makes it much easier for people unfamiliar with the code to contribute, plus easier to reason about in general.

# Changes
## `Composed`
From a user-perspective, the major change is that `∘` will now "flatten" the composition structure in the sense that
```julia
julia> using Bijectors; using Bijectors: Exp

julia> b = Exp()
Exp{0}()

julia> cb1 = b ∘ b;

julia> cb2 = b ∘ b;

julia> (cb1 ∘ cb2).ts # <= different
(Exp{0}(), Exp{0}(), Exp{0}(), Exp{0}())

julia> (cb1 ∘ cb2).ts isa NTuple{4, Exp{0}}
true
```
while in master it looks like
```julia
julia> using Bijectors; using Bijectors: Exp
[ Info: Recompiling stale cache file /home/tor/.julia/compiled/v1.1/Bijectors/39uFz.ji for Bijectors [76274a88-744f-5084-9051-94815aaf08c4]

julia> b = Exp()
Exp{0}()

julia> cb1 = b ∘ b;

julia> cb2 = b ∘ b;

julia> (cb1 ∘ cb2).ts # <= different
(Composed{Tuple{Exp{0},Exp{0}},0}((Exp{0}(), Exp{0}())), Composed{Tuple{Exp{0},Exp{0}},0}((Exp{0}(), Exp{0}())))

julia> (cb1 ∘ cb2).ts isa Tuple{Composed, Composed}
true
```
Note that this is _only_ the case for `∘`. In some cases it might be desirable to keep certain operations together, e.g. `Composed{Tuple{Scale, Shift}}}` represents an affine transformation. Therefore `composel` and `composer` will _not_ do this "flattening", but instead preserve the structure of the compositions.

The above is also described in the docstring of `Composed`.

### CODE REVIEW

File: src/bijectors/adbijector.jl [hunk 1]
```diff
@@ -0,0 +1,53 @@
+"""
+Abstract type for a `Bijector{N}` making use of auto-differentation (AD) to
+implement `jacobian` and, by impliciation, `logabsdetjac`.
+"""
+abstract type ADBijector{AD, N} <: Bijector{N} end
+
+# AD implementations
+function jacobian(b::ADBijector{<:ForwardDiffAD}, x::Real)
```
Comment (xukai92) [inline src/bijectors/adbijector.jl]: Can we merge the forward and inverse functions?

File: src/bijectors/adbijector.jl [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+end
+
+function jacobian(b::ADBijector{<:TrackerAD}, x::Real)
+    return Tracker.data(Tracker.gradient(b, x)[1])
+end
+function jacobian(b::Inversed{<:ADBijector{<:TrackerAD}}, y::Real)
+    return Tracker.data(Tracker.gradient(b, y)[1])
+end
+function jacobian(b::ADBijector{<:TrackerAD}, x::AbstractVector{<:Real})
+    # We extract `data` so that we don't returne a `Tracked` type
```
Comment (xukai92) [inline src/bijectors/adbijector.jl]: Typo. `returne` -> `return`

File: src/bijectors/scale.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
+struct Scale{T, N} <: Bijector{N}
+    a::T
+end
+
+Scale(a::T; dim::Val{D} = Val(0)) where {T<:Real, D} = Scale{T, D}(a)
+Scale(a::A; dim::Val{D} = Val(N)) where {T, D, N, A<:AbstractArray{T, N}} = Scale{A, D}(a)
+
+(b::Scale)(x) = b.a .* x
```
Comment (xukai92) [inline src/bijectors/scale.jl]: Is there a particular reason we sometimes use `@.` and sometimes do the dot operations explictly.

File: src/bijectors/truncated.jl [hunk 4]
```diff
@@ -0,0 +1,92 @@
+#######################################################
+# Constrained to unconstrained distribution bijectors #
+#######################################################
+struct TruncatedBijector{T} <: Bijector{0}
```
Comment (xukai92) [inline src/bijectors/truncated.jl]: Does the parameter here make things type stable?

**Comment by xukai92** [comment]
Looks good to me overall!

File: src/bijectors/adbijector.jl (hunk 1 above)
Comment (torfjelde) [reply src/bijectors/adbijector.jl]: You mean `Union{Inversed{B}, B}` where `B<:ADBijector{<:ForwardDiff}`? If so, that seems like a good idea to me:)

File: src/bijectors/scale.jl (hunk 3 above)
Comment (torfjelde) [reply src/bijectors/scale.jl]: I don't have a specific style that I follow for this. It's more a matter of how many dots I need to write out:)
I usually go with `@.` if I have more than 2 operations I want to fuse

File: src/bijectors/truncated.jl (hunk 4 above)
Comment (torfjelde) [reply src/bijectors/truncated.jl]: Depends on what the alternative is I suppose, but for example
```julia
struct TruncatedBijector <: Bijector{0}
    lb
    ub
end
```
would at least not be type-stable. But you could of course fix the types to be, say, `Float64`, but that seems unnecessary

File: src/bijectors/adbijector.jl (hunk 1 above)
Comment (xukai92) [reply src/bijectors/adbijector.jl]: Yes.

**Comment by xukai92** [comment]
@torfjelde Can I merge this now?

**Comment by torfjelde** [comment]
> @torfjelde Can I merge this now?

Yup! Though maybe I should bump the version number? I guess this would _technically_ be a breaking change since the circ-operation now has different behavior. Should I bump the minor- or patch-version?

File: src/bijectors/composed.jl [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    res = x
+    for b ∈ cb.ts
+        res = b(res)
+    end
+
+    return res
+end
+
+# recursive implementation like this allows type-inference
+_transform(x, b1::Bijector, b2::Bijector) = b2(b1(x))
```
Comment (torfjelde) [inline src/bijectors/composed.jl]: Implement this without recursion using `@generated` like we've done for `forward`.

**Comment by torfjelde** [comment]
Should be good now @kai!:)

**Comment by xukai92** [comment]
Thanks for the great work!

**Pull request merged**